File: javascript/2024/day04.js
import fs from 'node:fs';

const input = fs.readFileSync(0, 'utf-8');


let part1 = 0;
let part2 = 0;

let grid = [];

let row = 0;
let col = 0;

for (const line of input.split('\n')) {
    if (!line) {
        break;
    }
    col = 0;
    for (const c of line.split('')) {
        if (grid[row] === undefined) {
            grid[row] = [];
        }
        grid[row][col] = c;
        col++;
    }
    row++;
}

function getWord(len, getChar) {
    let letters = [];
    for (let i = 0; i < len; i++) {
        let c = getChar(i);

        if (c === undefined) {
            return undefined;
        }

        letters.push(c);
    }

    return letters.join('');
}

function gridAt(row, col) {
    if (grid[row] === undefined) {
        return undefined;
    }

    return grid[row][col];
}

function count(word, row, col) {
    let words = [];

    let c = gridAt(row, col);

    if (c !== 'X') {
        return 0;
    }

    const len = word.length;

    words.push(getWord(len, (i) => gridAt(row + i, col)));
    words.push(getWord(len, (i) => gridAt(row - i, col)));
    words.push(getWord(len, (i) => gridAt(row, col + i)));
    words.push(getWord(len, (i) => gridAt(row, col - i)));

    words.push(getWord(len, (i) => gridAt(row + i, col + i)));
    words.push(getWord(len, (i) => gridAt(row + i, col - i)));
    words.push(getWord(len, (i) => gridAt(row - i, col + i)));
    words.push(getWord(len, (i) => gridAt(row - i, col - i)));

    words = words.filter((w) => w === word);

    return words.length;
}

for (row = 0; row < grid.length; row++) {
    for (let col = 0; col < grid[row].length; col++) {
        part1 += count('XMAS', row, col);
    }
}

for (row = 0; row < grid.length; row++) {
    for (let col = 0; col < grid[row].length; col++) {
        let c = gridAt(row, col);

        if (c !== 'A') {
            continue;
        }

        let tl = gridAt(row - 1, col - 1);
        let tr = gridAt(row - 1, col + 1);
        let bl = gridAt(row + 1, col - 1);
        let br = gridAt(row + 1, col + 1);

        let w1 = [tl, br];
        let w2 = [tr, bl];

        w1 = w1.sort().join('');
        w2 = w2.sort().join('');

        if (w1 == 'MS' && w2 == 'MS') {
            part2++;
        }
    }
}

console.log(part1);
console.log(part2);

File: c/2024/day04.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 150
char grid[SIZE][SIZE];
int width = 0;
int height = 0;
const char xmas[] = "XMAS";

char charAt(int row, int col)
{
    if (row < 0 || row >= height || col < 0 || col >= width) {
        return 0;
    }

    return grid[row][col];
}

int isXmas(int row, int col, int dr, int dc)
{

    for (size_t i = 1; i < sizeof(xmas) - 1; i++) {
        row += dr;
        col += dc;
        char c = charAt(row, col);

        if (c != xmas[i]) {
            return 0;
        }
    }

    return 1;
}

int countXmasAt(int row, int col)
{
    int count = 0;

    // Check vertical, horizontal
    count += isXmas(row, col, -1, 0);
    count += isXmas(row, col, 1, 0);
    count += isXmas(row, col, 0, -1);
    count += isXmas(row, col, 0, 1);

    // Check diagonals
    count += isXmas(row, col, 1, 1);
    count += isXmas(row, col, 1, -1);
    count += isXmas(row, col, -1, -1);
    count += isXmas(row, col, -1, 1);

    return count;
}

int isCross(char a, char b)
{
    if (a == 'M' && b == 'S') {
        return 1;
    }

    if (a == 'S' && b == 'M') {
        return 1;
    }

    return 0;
}

int main()
{
    int part1 = 0;
    int part2 = 0;

    int row = 0;
    int col = 0;

    while (!feof(stdin)) {
        char c = fgetc(stdin);

        if (c == EOF) {
            break;
        }
        if (c == '\n') {
            width = col;
            col = 0;
            row++;
            if (row >= SIZE) {
                fprintf(stderr, "To many rows\n");
                return 1;
            }
            continue;
        }
        grid[row][col] = c;
        col++;

        if (col >= SIZE) {
            fprintf(stderr, "To many columns\n");
            return 1;
        }
    }

    height = row;

    for (row = 0; row < height; row++) {
        for (col = 0; col < width; col++) {
            char c = grid[row][col];

            if (c != 'X') {
                continue;
            }

            part1 += countXmasAt(row, col);
        }
    }

    for (row = 0; row < height; row++) {
        for (col = 0; col < width; col++) {
            char c = grid[row][col];

            if (c != 'A') {
                continue;
            }

            char tl = charAt(row - 1, col - 1);
            char tr = charAt(row - 1, col + 1);
            char bl = charAt(row + 1, col - 1);
            char br = charAt(row + 1, col + 1);

            if (tl && tr && bl && br == 0) {
                continue;
            }

            if (isCross(tl, br) && isCross(tr, bl)) {
                part2++;
            }
        }
    }

    printf("%d\n", part1);
    printf("%d\n", part2);

    return 0;
}
